staging/kos/pkg/controllers/subnet/validator.go

```diff
@@ -229,15 +230,27 @@ func NewValidationController(netIfc kosclientv1a1.NetworkV1alpha1Interface,
 			Help:      "Number of queue worker threads",
 		})
 
-	prometheus.MustRegister(subnetCreateToValidatedHistograms, subnetUpdateHistograms, liveListHistograms, liveListResultLengthHistogram, duplicateWorkCount, staleSubnetsSuppressionCount, cacheVsLiveSubnetMismatches, workerCount)
+	var eventRecorder k8seventrecord.EventRecorder
+
+	if !test {
+		// Register Prometheus metrics only if it's not a test because during a
+		// test the subnet validator constructor might be invoked multiple times
+		// and each invocation attempts to register the same Prometheus metrics
+		// names, which panics.
```

Comment:
Having a test flag in the constructor is ugly, so both suggestions were followed in 7ce6161 .

Concerning the `eventRecorder` as a constructor arg, I'm not sure whether you meant adding another constructor (used by the less general one) --- 7ce6161 doesn't do that, it simply modifies the existing constructor to accept the `eventRecorder` as a parameter (and removes the args whose sole purpose is setting up the `eventRecorder`). The real `eventRecorder` is created in the kos controller manager function which starts the validator (https://github.com/MikeSpreitzer/kube-examples/pull/83/files#diff-bbc277f7a59a5d8649a40e4007252ea2R102-R105); this is different wrt what some core K8s controllers do, i.e., creating the `eventRecorder` directly in the constructor and setting a fake one during tests simply by setting the appropriate field of the controller's struct. I chose the current approach because it was less code wrt to having more constructors, because it doesn't needlessly instantiate a real `eventRecorder` during tests, and because it leads to fewer constructor arguments.